sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -449,37 +449,109 @@ def test_spline_transformer_n_features_out(n_knots, include_bias, degree):
     assert splt.transform(X).shape[1] == splt.n_features_out_
 
 
-def test_polynomial_features():
-    # Test Polynomial Features
-    X1 = np.arange(6)[:, np.newaxis]
-    P1 = np.hstack([np.ones_like(X1),
-                    X1, X1 ** 2, X1 ** 3])
-    deg1 = 3
+@pytest.mark.parametrize(
+    "params, err_msg",
+    [
+        ({"degree": -1}, "degree must be a non-negative integer"),
+        ({"degree": 2.5}, "degree must be a non-negative int or tuple"),
+        ({"degree": "12"}, r"degree=\(min_degree, max_degree\) must"),
+        ({"degree": "string"}, "degree must be a non-negative int or tuple"),
+        ({"degree": (-1, 2)}, r"degree=\(min_degree, max_degree\) must"),
+        ({"degree": (0, 1.5)}, r"degree=\(min_degree, max_degree\) must"),
+        ({"degree": (3, 2)}, r"degree=\(min_degree, max_degree\) must"),
+    ],
+)
+def test_polynomial_features_input_validation(params, err_msg):
+    """Test that we raise errors for invalid input in PolynomialFeatures."""
+    X = [[1], [2]]
 
-    X2 = np.arange(6).reshape((3, 2))
-    x1 = X2[:, :1]
-    x2 = X2[:, 1:]
-    P2 = np.hstack([x1 ** 0 * x2 ** 0,
-                    x1 ** 1 * x2 ** 0,
-                    x1 ** 0 * x2 ** 1,
-                    x1 ** 2 * x2 ** 0,
-                    x1 ** 1 * x2 ** 1,
-                    x1 ** 0 * x2 ** 2])
-    deg2 = 2
+    with pytest.raises(ValueError, match=err_msg):
+        PolynomialFeatures(**params).fit(X)
+
+
+@pytest.fixture()
+def single_feature_degree2():
+    X = np.arange(6)[:, np.newaxis]
+    P = np.hstack([np.ones_like(X), X, X ** 2, X ** 3])
+    return X, P
 
-    for (deg, X, P) in [(deg1, X1, P1), (deg2, X2, P2)]:
-        P_test = PolynomialFeatures(deg, include_bias=True).fit_transform(X)
-        assert_array_almost_equal(P_test, P)
 
-        P_test = PolynomialFeatures(deg, include_bias=False).fit_transform(X)
-        assert_array_almost_equal(P_test, P[:, 1:])
+@pytest.mark.parametrize(
+    "degree, include_bias, interaction_only, indices",
+    [
+        (3, True, False, slice(None, None)),
+        (3, False, False, slice(1, None)),
+        (3, True, True, [0, 1]),
+        (3, False, True, [1]),
+        ((2, 3), True, False, [0, 2, 3]),
+        ((2, 3), False, False, [2, 3]),
+        ((2, 3), True, True, [0]),
+        ((2, 3), False, True, []),
+    ]
+)
+def test_polynomial_features_degree_3(
+    single_feature_degree2,
+    degree,
+    include_bias,
+    interaction_only,
+    indices
+):
+    """Test Polynomial Features of degree 3."""
+    X, P = single_feature_degree2
+    tf = PolynomialFeatures(
+        degree=degree,
+        include_bias=include_bias,
+        interaction_only=interaction_only
+    ).fit(X)
+    assert_allclose(tf.transform(X), P[:, indices])
+    if tf.n_output_features_ > 0:
+        assert tf.powers_.shape == (tf.n_output_features_, tf.n_features_in_)
+
+
+@pytest.fixture()
+def two_feature_degree3():
```

Comment:
typo?

```suggestion
def two_feature_degree2():
```

Although it would be nice to have a similar fixture for degree3.

Reply:
Done. Now we have the 2 features case up to degree 3.